### PR TITLE
thoughts?

### PR DESCRIPTION
**Pull request by safareli**
I started this off the err branch, and target is that same branch to make diff nicer:

This snippet demonstrates how things play together:

```hs
app :: App (Free DSL)
interpret :: Config -> Free DSL ~> IO
mkConfig :: IO Config

main = do
  conf <- mkConfig appData
  serveApp $ transformApp (interpret conf) app
```

### CODE REVIEW

File: src/Network/ABCI.hs [hunk 1]
```diff
@@ -21,74 +22,47 @@ import           Network.ABCI.Types.App               (App (..))
...
-  app <- mkApp (appSockAddr appData)
-  runConduit (setupConduit app appData)
+  :: ServerSettings
+  -> (AppData -> IO (AppRunner IO))
+  -> IO ()
+serveAppWith cfg mkRunner = runTCPServer cfg $ \appData -> do
+  runner <- mkRunner appData
+  runConduit $
+      appSource appData
+    .| CL.mapM runner
```
Comment (safareli) [inline src/Network/ABCI.hs:39]: We just use one mapM now as all other transformations were bunch of maps and one mapM and they can be fused together.

File: src/Network/ABCI.hs [hunk 2]
```diff
@@ -21,74 +22,47 @@ import           Network.ABCI.Types.App               (App (..))
...
-     appSource appData
-  .| Wire.decodeLengthPrefixC
-  .| CL.map (traverse decodeRequestValue =<<)
-  .| CL.mapM (respondWith app)
-  .| CL.map (map PL.encodeMessage)
-  .| Wire.encodeLengthPrefixC
-  .| appSink appData
+runApp :: Monad m => App m -> AppRunner m
+runApp (App app) bsInput = do
+  responcesE <- runExceptT $ do
```
Comment (safareli) [inline src/Network/ABCI.hs:49]: ExceptT is used so we don have functions like `Either Error x -> ....`

File: src/Network/ABCI.hs [hunk 3]
```diff
@@ -21,74 +22,47 @@ import           Network.ABCI.Types.App               (App (..))
...
-  -> ConduitT i o m ()
-setupConduit app appData =
-     appSource appData
-  .| Wire.decodeLengthPrefixC
-  .| CL.map (traverse decodeRequestValue =<<)
-  .| CL.mapM (respondWith app)
-  .| CL.map (map PL.encodeMessage)
-  .| Wire.encodeLengthPrefixC
-  .| appSink appData
+runApp :: Monad m => App m -> AppRunner m
```
Comment (safareli) [inline src/Network/ABCI.hs:47]: it's pure function and can make testing easier

**Comment by safareli** [comment]
This is good point https://github.com/f-o-a-m/hs-abci-server/pull/24#discussion_r306934838 so removed the AppData all together as for testing one can just use `runApp`

File: src/Network/ABCI.hs [hunk 4]
```diff
@@ -21,74 +22,41 @@ import           Network.ABCI.Types.App               (App (..))
...
```
Comment (safareli) [inline src/Network/ABCI.hs:47]: With current arrangement I would say to even move runApp and decodeLengthPrefix/encodeLengthPrefix to App module

**Comment by martyall** [comment]
(thoughts given on rocket chat)

**Comment by safareli** [comment]
#28

**Pull request closed without merging**